File: packages/video-contracts/src/time.ts
import { z } from "zod";

/**
 * Exact Rational Time representation.
 * Eliminates floating-point rounding errors and audio/video desynchronization.
 */
export const RationalTimeSchema = z.object({
  value: z.number().int().describe("Numerator in timescale units"),
  timescale: z.number().int().positive().describe("Units per second (e.g. 48000 for audio, 60000 for 60fps)"),
});

export type RationalTime = z.infer<typeof RationalTimeSchema>;

export const TimeRangeSchema = z.object({
  start: RationalTimeSchema,
  duration: RationalTimeSchema,
});

export type TimeRange = z.infer<typeof TimeRangeSchema>;

export const FrameRateSchema = z.object({
  numerator: z.number().int().positive(),
  denominator: z.number().int().positive(),
});

export type FrameRate = z.infer<typeof FrameRateSchema>;

/**
 * Rational arithmetic utilities
 */
export class RationalTimeMath {
  /**
   * Converts a float seconds timestamp to exact RationalTime
   */
  static fromSeconds(seconds: number, timescale: number = 48000): RationalTime {
    return {
      value: Math.round(seconds * timescale),
      timescale,
    };
  }

  /**
   * Converts RationalTime to floating-point seconds
   */
  static toSeconds(time: RationalTime): number {
    if (time.timescale === 0) return 0;
    return time.value / time.timescale;
  }

  /**
   * Converts frame index at specific frame rate to exact RationalTime without float rounding
   */
  static fromFrames(frameIndex: number, fps: FrameRate, targetTimescale?: number): RationalTime {
    // Exact fractional representation: (frameIndex * fps.denominator) / fps.numerator
    const rawTime: RationalTime = {
      value: frameIndex * fps.denominator,
      timescale: fps.numerator,
    };

    if (targetTimescale && targetTimescale !== fps.numerator) {
      const common = this.lcm(rawTime.timescale, targetTimescale);
      const scaledVal = rawTime.value * (common / rawTime.timescale);
      return this.simplify({ value: scaledVal, timescale: common });
    }

    return this.simplify(rawTime);
  }

  /**
   * Converts RationalTime to frame number using exact integer division
   */
  static toFrames(time: RationalTime, fps: FrameRate): number {
    if (time.timescale === 0 || fps.denominator === 0) return 0;
    const num = BigInt(time.value) * BigInt(fps.numerator);
    const den = BigInt(time.timescale) * BigInt(fps.denominator);
    if (den === 0n) return 0;
    return Number(num / den);
  }

  /**
   * Adds two RationalTimes, finding common timescale
   */
  static add(a: RationalTime, b: RationalTime): RationalTime {
    if (a.timescale === b.timescale) {
      return { value: a.value + b.value, timescale: a.timescale };
    }
    const commonScale = this.lcm(a.timescale, b.timescale);
    const valA = a.value * (commonScale / a.timescale);
    const valB = b.value * (commonScale / b.timescale);
    return this.simplify({ value: valA + valB, timescale: commonScale });
  }

  /**
   * Subtracts b from a: a - b
   */
  static sub(a: RationalTime, b: RationalTime): RationalTime {
    if (a.timescale === b.timescale) {
      return { value: a.value - b.value, timescale: a.timescale };
    }
    const commonScale = this.lcm(a.timescale, b.timescale);
    const valA = a.value * (commonScale / a.timescale);
    const valB = b.value * (commonScale / b.timescale);
    return this.simplify({ value: valA - valB, timescale: commonScale });
  }

  /**
   * Multiplies RationalTime by a numeric factor
   */
  static mul(time: RationalTime, factor: number): RationalTime {
    return this.simplify({
      value: Math.round(time.value * factor),
      timescale: time.timescale,
    });
  }

  /**
   * Compares two RationalTimes: -1 if a < b, 0 if a === b, 1 if a > b
   */
  static compare(a: RationalTime, b: RationalTime): number {
    const diff = this.sub(a, b);
    if (diff.value === 0) return 0;
    return diff.value > 0 ? 1 : -1;
  }

  /**
   * Clamps RationalTime between min and max
   */
  static clamp(time: RationalTime, min: RationalTime, max: RationalTime): RationalTime {
    if (this.compare(time, min) < 0) return min;
    if (this.compare(time, max) > 0) return max;
    return time;
  }

  /**
   * Checks if RationalTime is zero
   */
  static isZero(time: RationalTime): boolean {
    return time.value === 0;
  }

  /**
   * Simplifies rational fraction using GCD
   */
  static simplify(time: RationalTime): RationalTime {
    if (time.value === 0) {
      return { value: 0, timescale: time.timescale || 48000 };
    }
    const divisor = this.gcd(Math.abs(time.value), time.timescale);
    return {
      value: time.value / divisor,
      timescale: time.timescale / divisor,
    };
  }

  private static gcd(a: number, b: number): number {
    return b === 0 ? a : this.gcd(b, a % b);
  }

  private static lcm(a: number, b: number): number {
    if (a === 0 || b === 0) return 0;
    return Math.abs((a * b) / this.gcd(a, b));
  }
}

File: native/video-engine-core/crates/engine-timeline/src/time.rs
use serde::{Deserialize, Serialize};
use std::ops::{Add, Sub};

/// Exact rational time representation to prevent audio/video drift.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct RationalTime {
    pub value: i64,
    pub timescale: u32,
}

impl RationalTime {
    pub const ZERO: Self = Self { value: 0, timescale: 48000 };

    pub fn new(value: i64, timescale: u32) -> Self {
        assert!(timescale > 0, "Timescale must be strictly positive");
        Self::simplify(value, timescale)
    }

    pub fn from_seconds(seconds: f64, timescale: u32) -> Self {
        let value = (seconds * timescale as f64).round() as i64;
        Self::new(value, timescale)
    }

    pub fn to_seconds(&self) -> f64 {
        if self.timescale == 0 {
            0.0
        } else {
            self.value as f64 / self.timescale as f64
        }
    }

    pub fn from_frames(frame_index: i64, fps_num: u32, fps_den: u32, timescale: u32) -> Self {
        let seconds = (frame_index as f64 * fps_den as f64) / fps_num as f64;
        Self::from_seconds(seconds, timescale)
    }

    pub fn to_frames(&self, fps_num: u32, fps_den: u32) -> i64 {
        let seconds = self.to_seconds();
        ((seconds * fps_num as f64) / fps_den as f64).floor() as i64
    }

    fn simplify(mut value: i64, mut timescale: u32) -> Self {
        let divisor = gcd(value.unsigned_abs(), timescale as u64);
        if divisor > 1 {
            value /= divisor as i64;
            timescale /= divisor as u32;
        }
        Self { value, timescale }
    }
}

impl Add for RationalTime {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        if self.timescale == rhs.timescale {
            return Self::new(self.value + rhs.value, self.timescale);
        }
        let common = lcm(self.timescale as u64, rhs.timescale as u64) as u32;
        let v1 = self.value * (common / self.timescale) as i64;
        let v2 = rhs.value * (common / rhs.timescale) as i64;
        Self::new(v1 + v2, common)
    }
}

impl Sub for RationalTime {
    type Output = Self;

    fn sub(self, rhs: Self) -> Self::Output {
        if self.timescale == rhs.timescale {
            return Self::new(self.value - rhs.value, self.timescale);
        }
        let common = lcm(self.timescale as u64, rhs.timescale as u64) as u32;
        let v1 = self.value * (common / self.timescale) as i64;
        let v2 = rhs.value * (common / rhs.timescale) as i64;
        Self::new(v1 - v2, common)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct TimeRange {
    pub start: RationalTime,
    pub duration: RationalTime,
}

impl TimeRange {
    pub fn new(start: RationalTime, duration: RationalTime) -> Self {
        Self { start, duration }
    }

    pub fn end(&self) -> RationalTime {
        self.start + self.duration
    }

    pub fn contains(&self, time: RationalTime) -> bool {
        time.to_seconds() >= self.start.to_seconds() && time.to_seconds() <= self.end().to_seconds()
    }
}

fn gcd(mut a: u64, mut b: u64) -> u64 {
    while b != 0 {
        let temp = b;
        b = a % b;
        a = temp;
    }
    a
}

fn lcm(a: u64, b: u64) -> u64 {
    if a == 0 || b == 0 {
        0
    } else {
        (a * b) / gcd(a, b)
    }
}
